src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -2237,9 +2239,27 @@ private static IEnumerable<T> ForceLazy<T>(IEnumerable<T> source)
         /// <returns>The underlying array.</returns>
         private static T[] GetUnderlyingArray<T>(ImmutableArray<T> array)
         {
-            FieldInfo arrayField = typeof(ImmutableArray<T>)
-                .GetField("array", BindingFlags.Instance | BindingFlags.NonPublic);
-            return (T[])arrayField.GetValue(array);
+            //
+            // There is no documented way of doing this so this helper is inherently fragile.
+            //
+            // The prior version of this used Reflection to get at the private "array" field directly. This will not work on .NET Native
+            // due to the prohibition on internal framework Reflection.
+            //
+            // This alternate method is despicable but ImmutableArray`1 has a documented contract of being exactly one reference-type field in size
+            // (for example, ImmutableInterlocked depends on it.) 
+            //
+            // That leaves precious few candidates for which field is the "array" field...
+            //
+            T[] underlyingArray = Unsafe.As<ImmutableArray<T>, ImmutableArrayLayout<T>>(ref array).array;
+            if (underlyingArray != null && !(((object)underlyingArray) is T[]))
```

Comment:
ImmutableArray<T> is a struct so a null array has to be considered a plausible possibility. In any case, there's no danger of heap corruption in the null case which is what this guard is really trying to avoid.